# IS 362 Assignment 5

## Prompt:

Your task in this week’s assignment is to answer three questions:
1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather?

Being successful on this assignment requires using both:
* Your developing skills in Python and pandas, and
* Your research skills in validating your candidate answers in context

Tasks
* Use the information in the .CSV files found at
https://github.com/hadley/nycflights13/tree/master/data-raw for your source data.
* Write your code using Python and pandas, and deliver it in a Jupyter notebook. You should post your notebook in a GitHub repository, and send me a link to the notebook.
* For each of the three questions listed above, you’ll need to verify your answers (in two cases by searching for additional confirming information, and in the third case being alert for possible outliers). For the first two questions above, you may want to start by generating a list of the five
northernmost or easternmost U.S. airports.
* In addition to providing code, you should state your answers to each of the above questions (in Markdown) in the Jupyter notebook.
* To receive full credit, you’ll need to include in your Jupyter notebook both your code and your documented assumptions and any changes that you made to the data based on your research (using Markdown) in your Jupyter notebook.

## Question 1

What is the northernmost airport in the United States?

In [1]:
# import pandas
import pandas as pd

In [2]:
# read the airports.csv and save as a pandas DataFrame
airport_df = pd.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/airports.csv')

In [3]:
# there are seven columns
# column 'lat' holds latitude values
airport_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458 entries, 0 to 1457
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   faa     1458 non-null   object 
 1   name    1458 non-null   object 
 2   lat     1458 non-null   float64
 3   lon     1458 non-null   float64
 4   alt     1458 non-null   int64  
 5   tz      1458 non-null   int64  
 6   dst     1458 non-null   object 
 7   tzone   1455 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 91.2+ KB


In [4]:
# .head() shows a snapshot of the data
airport_df.head()

faa                           name        lat        lon   alt  tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044  -5   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264  -6   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801  -6   A   
3  06N                Randall Airport  41.431912 -74.391561   523  -5   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11  -5   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

In [5]:
# .sort_values() will sort the DataFrame by the specified column, in this case latitude
# the latitude range of this data set is 19.721375 (ITO airport in Hawaii) at its most southern point
# up to 72.270833 (EEN airport in New Hampshire) at its most northern point
airport_df.sort_values('lat')

faa                                        name        lat         lon  \
679   ITO                                   Hilo Intl  19.721375 -155.048469   
734   KOA                        Kona Intl At Keahole  19.738767 -156.045631   
231   BSF                                Bradshaw Aaf  19.760056 -155.553717   
1403  WKL                           Waikoloa Heliport  19.913600 -155.864000   
930   MUE                               Waimea Kohala  20.001328 -155.668108   
...   ...                                         ...        ...         ...   
152   ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   
708   K03                               Wainwright As  70.613378 -159.860350   
110   AIN                          Wainwright Airport  70.638056 -159.994722   
230   BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
417   EEN                     Dillant Hopkins Airport  72.270833   42.898333   

       alt  tz dst              tzone  
679     38 -10   N   Pacific/Honolulu  
734     47 -10   N   Pacific/Honolulu  
231   6190 -10   A   Pacific/Honolulu  
1403   109 -10   N   Pacific/Honolulu  
930   2671 -10   A   Pacific/Honolulu  
...    ...  ..  ..                ...  
152     96  -9   A  America/Anchorage  
708     35  -9   A  America/Anchorage  
110     41  -9   A  America/Anchorage  
230     44  -9   A  America/Anchorage  
417    149  -5   A                NaN  

[1458 rows x 8 columns]

I'm surprised to see that the northernmost airport is in New Hampshire, as I am sure there are airports in the states north of NH. After looking at the listed latitude and longitude values of EEN airport in Google Maps, I see that the latitude of EEN is incorrect. The correct latitude should be (what is shown as its longitude) 42.898333, and its latitude should be -72.270833.

In [6]:
# correct values
airport_df.loc[airport_df['faa']=='EEN', 'lat'] = 42.898333
airport_df.loc[airport_df['faa']=='EEN', 'lon'] = -72.270833
airport_df.sort_values('lat')

faa                                        name        lat         lon  \
679   ITO                                   Hilo Intl  19.721375 -155.048469   
734   KOA                        Kona Intl At Keahole  19.738767 -156.045631   
231   BSF                                Bradshaw Aaf  19.760056 -155.553717   
1403  WKL                           Waikoloa Heliport  19.913600 -155.864000   
930   MUE                               Waimea Kohala  20.001328 -155.668108   
...   ...                                         ...        ...         ...   
1363  UUK                        Ugnu-Kuparuk Airport  70.330833 -149.597500   
152   ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   
708   K03                               Wainwright As  70.613378 -159.860350   
110   AIN                          Wainwright Airport  70.638056 -159.994722   
230   BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   

       alt  tz dst              tzone  
679     38 -10   N   Pacific/Honolulu  
734     47 -10   N   Pacific/Honolulu  
231   6190 -10   A   Pacific/Honolulu  
1403   109 -10   N   Pacific/Honolulu  
930   2671 -10   A   Pacific/Honolulu  
...    ...  ..  ..                ...  
1363    67  -9   A  America/Anchorage  
152     96  -9   A  America/Anchorage  
708     35  -9   A  America/Anchorage  
110     41  -9   A  America/Anchorage  
230     44  -9   A  America/Anchorage  

[1458 rows x 8 columns]

In [7]:
northernmost_airport = airport_df[airport_df.lat == airport_df.lat.max()]
northernmost_airport

faa                        name        lat         lon  alt  tz dst  \
230  BRW  Wiley Post Will Rogers Mem  71.285446 -156.766003   44  -9   A   

                 tzone  
230  America/Anchorage

Using Google Maps as a reference, I verified that BRW airport in Alaska appears to be the at the northernmost point of this data set.

In [8]:
northernmost_airport_name = (northernmost_airport['name']).to_string(index=False).lstrip()
northernmost_airport_name

'Wiley Post Will Rogers Mem'

## Question 1 Answer

In [9]:
print('The northernmost airport in the United States is {}.'.format(northernmost_airport_name))

The northernmost airport in the United States is Wiley Post Will Rogers Mem.


## Question 2

What is the easternmost airport in the United States?

In [10]:
# using the same dataset, the 'lon' column holds longtitude values
airport_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458 entries, 0 to 1457
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   faa     1458 non-null   object 
 1   name    1458 non-null   object 
 2   lat     1458 non-null   float64
 3   lon     1458 non-null   float64
 4   alt     1458 non-null   int64  
 5   tz      1458 non-null   int64  
 6   dst     1458 non-null   object 
 7   tzone   1455 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 91.2+ KB


In [11]:
# view the range of the lon values
airport_df.lon.sort_values()

95     -176.646000
112    -174.206389
522    -171.733000
1281   -170.493000
1251   -170.220444
           ...    
624     -67.792056
444     -67.012694
396     112.457000
942     117.759000
1290    174.113620
Name: lon, Length: 1458, dtype: float64

In [12]:
# grouping by timezone may narrow the search
airport_df['tzone'].value_counts()

America/New_York       519
America/Chicago        342
America/Anchorage      239
America/Los_Angeles    176
America/Denver         119
America/Phoenix         38
Pacific/Honolulu        18
America/Vancouver        2
Asia/Chongqing           2
Name: tzone, dtype: int64

In [13]:
# the corresponding 'tz' value may help identify the easternmost tzone
airport_df.sort_values(['tz', 'tzone'])

faa                                  name        lat         lon   alt  \
207   BKH                    Barking Sands Pmrf  22.022833 -159.785000    23   
231   BSF                          Bradshaw Aaf  19.760056 -155.553717  6190   
580   HDH                            Dillingham  21.579475 -158.197281    14   
587   HHI                           Wheeler Aaf  21.483500 -158.039667   837   
600   HNL                         Honolulu Intl  21.318681 -157.922428    13   
...   ...                                   ...        ...         ...   ...   
1457  ZYP                          Penn Station  40.750500  -73.993500    35   
417   EEN               Dillant Hopkins Airport  42.898333  -72.270833   149   
815   LRO  Mount Pleasant Regional-Faison Field  32.538700  -79.469700    12   
396   DVT         Deer Valley Municipal Airport  33.411700  112.457000  1478   
942   MYF                      Montgomery Field  32.475900  117.759000    17   

      tz dst             tzone  
207  -10   A  Pacific/Honolulu  
231  -10   A  Pacific/Honolulu  
580  -10   A  Pacific/Honolulu  
587  -10   A  Pacific/Honolulu  
600  -10   N  Pacific/Honolulu  
...   ..  ..               ...  
1457  -5   A  America/New_York  
417   -5   A               NaN  
815   -5   A               NaN  
396    8   A    Asia/Chongqing  
942    8   A    Asia/Chongqing  

[1458 rows x 8 columns]

In [14]:
# which tzones correspond with which tz values?
by_tzone = airport_df.groupby('tzone').tz.agg(['min', 'max'])
by_tzone

min  max
tzone                        
America/Anchorage     -9   -9
America/Chicago       -6   -6
America/Denver        -7   -7
America/Los_Angeles   -8   -8
America/New_York      -5   -5
America/Phoenix       -7   -7
America/Vancouver     -8   -8
Asia/Chongqing         8    8
Pacific/Honolulu     -10  -10

In [15]:
# tz -5 appears to be the easternmost tz in the United States
by_tzone.sort_values('min')

min  max
tzone                        
Pacific/Honolulu     -10  -10
America/Anchorage     -9   -9
America/Los_Angeles   -8   -8
America/Vancouver     -8   -8
America/Denver        -7   -7
America/Phoenix       -7   -7
America/Chicago       -6   -6
America/New_York      -5   -5
Asia/Chongqing         8    8

In [16]:
# sorting the df by lon values in the -5 tz narrows the search
tz_neg5 = airport_df[(airport_df.tz == -5)].sort_values('lon')
tz_neg5

faa                                 name        lat        lon   alt  \
310   CMX     Houghton County Memorial Airport  47.168400 -88.489100  1095   
913   MQT         Sawyer International Airport  46.353611 -87.395278  1221   
623   HUF          Terre Haute Intl Hulman Fld  39.451464 -87.307561   589   
449   ESC                 Delta County Airport  45.722778 -87.093611   609   
763   LAF            Purude University Airport  40.412306 -86.936889   606   
...   ...                                  ...        ...        ...   ...   
1398  WFK  Northern Aroostook Regional Airport  47.285556 -68.312778   988   
1101  PQI  Northern Maine Rgnl At Presque Isle  46.688958 -68.044797   534   
259   CAR                         Caribou Muni  46.871500 -68.017917   626   
624   HUL                         Houlton Intl  46.123083 -67.792056   489   
444   EPM           Eastport Municipal Airport  44.910111 -67.012694    45   

      tz dst             tzone  
310   -5   A  America/New_York  
913   -5   A  America/New_York  
623   -5   A  America/New_York  
449   -5   A  America/New_York  
763   -5   A  America/New_York  
...   ..  ..               ...  
1398  -5   A  America/New_York  
1101  -5   A  America/New_York  
259   -5   A  America/New_York  
624   -5   A  America/New_York  
444   -5   A  America/New_York  

[521 rows x 8 columns]

In [17]:
easternnmost_airport = tz_neg5[tz_neg5.lon == tz_neg5.lon.max()]
easternnmost_airport

faa                        name        lat        lon  alt  tz dst  \
444  EPM  Eastport Municipal Airport  44.910111 -67.012694   45  -5   A   

                tzone  
444  America/New_York

Using Google Maps as a reference, I verified that EPM airport in Maine appears to be the at the easternmost point of this data set.

In [18]:
easternmost_airport_name = (easternnmost_airport['name']).to_string(index=False).lstrip()
easternmost_airport_name

'Eastport Municipal Airport'

## Question 2 Answer

In [19]:
print('The easternmost airport in the United States is {}.'.format(easternmost_airport_name))

The easternmost airport in the United States is Eastport Municipal Airport.


## Question 3

On February 12th, 2013, which New York area airport had the windiest weather?

In [20]:
# read the weather.csv and save as a pandas DataFrame
weather_df = pd.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/weather.csv')

In [21]:
# there are 15 columns, three of which pertain to wind
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26115 entries, 0 to 26114
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   origin      26115 non-null  object 
 1   year        26115 non-null  int64  
 2   month       26115 non-null  int64  
 3   day         26115 non-null  int64  
 4   hour        26115 non-null  int64  
 5   temp        26114 non-null  float64
 6   dewp        26114 non-null  float64
 7   humid       26114 non-null  float64
 8   wind_dir    25655 non-null  float64
 9   wind_speed  26111 non-null  float64
 10  wind_gust   5337 non-null   float64
 11  precip      26115 non-null  float64
 12  pressure    23386 non-null  float64
 13  visib       26115 non-null  float64
 14  time_hour   26115 non-null  object 
dtypes: float64(9), int64(4), object(2)
memory usage: 3.0+ MB


In [22]:
# .head() shows a snapshot of the data
weather_df.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

In [23]:
# in the .head() snapshot, the wind_gust column appears hold only null values
# I can verify the number of null values using the .isnull() and .sum() methods
weather_df.isnull().sum()

origin            0
year              0
month             0
day               0
hour              0
temp              1
dewp              1
humid             1
wind_dir        460
wind_speed        4
wind_gust     20778
precip            0
pressure       2729
visib             0
time_hour         0
dtype: int64

More than 20 thousand of the approximately 26 thousand records have null wind_gust values. 460 wind_dir values are null. To answer question 3 I will use the wind_speed values. The highest wind_speed values will be considered the windiest.

In [24]:
# the question asks about NY area airports
# which airports are included in this data set?
weather_df.origin.value_counts()

LGA    8706
JFK    8706
EWR    8703
Name: origin, dtype: int64

In [25]:
# filtering by date and airport returns the wind_speed per hour
ewr_on_feb122013 = weather_df[(weather_df.month == 2) & (weather_df.day == 12) & (weather_df.year == 2013) & (weather_df.origin == 'EWR')].sort_values('hour')
ewr_on_feb122013

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
1006    EWR  2013      2   12     0  39.92  39.02  96.55     240.0   
1007    EWR  2013      2   12     1  39.92  37.94  92.56     250.0   
1008    EWR  2013      2   12     2  39.92  28.04  62.21     270.0   
1009    EWR  2013      2   12     3  39.02  26.96  61.63     260.0   
1010    EWR  2013      2   12     4  39.02  26.96  64.29     280.0   
1011    EWR  2013      2   12     5  39.92  26.96  60.29     270.0   
1012    EWR  2013      2   12     6  39.02  26.96  61.63     270.0   
1013    EWR  2013      2   12     7  39.92  24.98  54.81     260.0   
1014    EWR  2013      2   12     8  42.08  26.60  60.29     250.0   
1015    EWR  2013      2   12     9  41.90  24.98  50.77     290.0   
1016    EWR  2013      2   12    10  42.08  26.06  52.73     260.0   
1017    EWR  2013      2   12    11  42.98  24.98  48.70     260.0   
1018    EWR  2013      2   12    12  44.06  26.06  48.87     270.0   
1019    EWR  2013      2   12    13  46.04  26.06  45.33     280.0   
1020    EWR  2013      2   12    14  44.96  26.06  47.23     280.0   
1021    EWR  2013      2   12    15  42.98  24.98  48.70     300.0   
1022    EWR  2013      2   12    16  42.98  24.08  46.91     270.0   
1023    EWR  2013      2   12    17  41.00  24.08  50.63     260.0   
1024    EWR  2013      2   12    18  39.92  24.08  52.79     270.0   
1025    EWR  2013      2   12    19  39.92  24.08  52.79     270.0   
1026    EWR  2013      2   12    20  37.94  24.08  57.04     280.0   
1027    EWR  2013      2   12    21  35.96  24.08  61.67     300.0   
1028    EWR  2013      2   12    22  35.96  23.00  58.94     260.0   
1029    EWR  2013      2   12    23  35.96  23.00  58.94     250.0   

      wind_speed  wind_gust  precip  pressure  visib             time_hour  
1006     6.90468        NaN     0.0    1006.9   10.0  2013-02-12T05:00:00Z  
1007     9.20624        NaN     0.0    1007.2   10.0  2013-02-12T06:00:00Z  
1008    20.71404   25.31716     0.0    1007.8   10.0  2013-02-12T07:00:00Z  
1009  1048.36058        NaN     0.0    1008.3   10.0  2013-02-12T08:00:00Z  
1010    12.65858        NaN     0.0    1008.8   10.0  2013-02-12T09:00:00Z  
1011    13.80936        NaN     0.0       NaN   10.0  2013-02-12T10:00:00Z  
1012    16.11092        NaN     0.0    1009.8   10.0  2013-02-12T11:00:00Z  
1013    14.96014   20.71404     0.0    1010.7   10.0  2013-02-12T12:00:00Z  
1014    17.26170        NaN     0.0       NaN   10.0  2013-02-12T13:00:00Z  
1015    12.65858        NaN     0.0    1011.5   10.0  2013-02-12T14:00:00Z  
1016    17.26170   23.01560     0.0    1012.0   10.0  2013-02-12T15:00:00Z  
1017    17.26170        NaN     0.0    1012.3   10.0  2013-02-12T16:00:00Z  
1018    21.86482   31.07106     0.0    1012.5   10.0  2013-02-12T17:00:00Z  
1019    13.80936   31.07106     0.0    1012.3   10.0  2013-02-12T18:00:00Z  
1020    17.26170   24.16638     0.0    1011.9   10.0  2013-02-12T19:00:00Z  
1021    12.65858   19.56326     0.0    1011.8   10.0  2013-02-12T20:00:00Z  
1022    13.80936        NaN     0.0    1011.3   10.0  2013-02-12T21:00:00Z  
1023    16.11092   26.46794     0.0    1013.8   10.0  2013-02-12T22:00:00Z  
1024    11.50780        NaN     0.0    1014.6   10.0  2013-02-12T23:00:00Z  
1025    12.65858        NaN     0.0    1015.0   10.0  2013-02-13T00:00:00Z  
1026     9.20624        NaN     0.0    1013.9   10.0  2013-02-13T01:00:00Z  
1027     5.75390        NaN     0.0    1013.0   10.0  2013-02-13T02:00:00Z  
1028     6.90468        NaN     0.0    1013.4   10.0  2013-02-13T03:00:00Z  
1029     4.60312        NaN     0.0    1013.7   10.0  2013-02-13T04:00:00Z

In [26]:
# the mean wind_speed can be calculated on the filtered df, however it is inefficient to perform this task on every airport in this df
ewr_on_feb122013.wind_speed.mean()

56.38822

In [27]:
# filter by date, include all airports
on_feb122013 = weather_df[(weather_df.month == 2) & (weather_df.day == 12) & (weather_df.year == 2013)]

In [28]:
# calculate the mean wind_speed on of all hours in the day for each airport
average_wind_speed_on_feb122013 = on_feb122013.groupby('origin').wind_speed.mean()
average_wind_speed_on_feb122013

origin
EWR    56.38822
JFK    14.38475
LGA    14.96014
Name: wind_speed, dtype: float64

In [29]:
# identify the max average wind_speed on the specified day
average_wind_speed_on_feb122013.max()

56.38822000000001

In [30]:
# identify the airport name
windiest_airport = average_wind_speed_on_feb122013[average_wind_speed_on_feb122013 == average_wind_speed_on_feb122013.max()]
windiest_airport

origin
EWR    56.38822
Name: wind_speed, dtype: float64

In [31]:
windiest_airport_name = windiest_airport.index[0]
windiest_airport_name

'EWR'

## Question 3 Answer

In [32]:
print('The windiest New York area airport on February 12, 2013 was {}.'.format(windiest_airport_name))

The windiest New York area airport on February 12, 2013 was EWR.
